File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        const int N = nums.size();
        vector<int> dp(N);
        int maxi = 1;
        dp[N - 1] = 1;

        for (int i = N - 2; i >= 0; i--) {
            dp[i] = 1;
            int curr = nums[i];

            for (int j = i + 1; j < N; j++) {
                if (curr < nums[j]){ dp[i] = max(dp[i], 1 + dp[j]); }
            }
            maxi = max(maxi, dp[i]);
        }

        return maxi;

    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        
        N = len(nums)
        dp = [1] * N
        ans = 1
        
        for i in range(N-2, -1, -1):
            curr = nums[i]
            for j in range(i+1, N):
                if curr < nums[j]:
                    dp[i] = max(dp[i], 1 + dp[j])
            ans = max(ans, dp[i])
            
        return ans
        
        
        